// azure-pipelines-tasks-kubernetes-common/helmutility.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStableHelmVersion = exports.downloadHelm = exports.getHelm = void 0;
const path = require("path");
const fs = require("fs");
const toolLib = require("azure-pipelines-tool-lib/tool");
const os = require("os");
const util = require("util");
const uuidV4 = require("uuid/v4");
const tl = require("azure-pipelines-task-lib/task");
const utility_1 = require("./utility");
const osutil = require("./osutility");
const semver = require('semver');
tl.setResourcePath(path.join(__dirname, 'module.json'), true);
const helmToolName = 'helm';
const helmAllReleasesUrl = 'https://api.github.com/repos/helm/helm/releases';
const stableHelmVersion = 'v3.1.2';
function getHelm(version) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            return Promise.resolve(tl.which('helm', true));
        }
        catch (ex) {
            return downloadHelm(version);
        }
    });
}
exports.getHelm = getHelm;
function downloadHelm(version) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!version) {
            version = yield getStableHelmVersion();
        }
        let cachedToolpath = toolLib.findLocalTool(helmToolName, version);
        if (!cachedToolpath) {
            let helmDownloadPath;
            try {
                helmDownloadPath = yield toolLib.downloadTool(getHelmDownloadURL(version), helmToolName + '-' + version + '-' + uuidV4() + '.zip');
            }
            catch (exception) {
                throw new Error(tl.loc('HelmDownloadFailed', getHelmDownloadURL(version), exception));
            }
            const unzipedHelmPath = yield toolLib.extractZip(helmDownloadPath);
            cachedToolpath = yield toolLib.cacheDir(unzipedHelmPath, helmToolName, version);
        }
        const helmpath = findHelm(cachedToolpath);
        if (!helmpath) {
            throw new Error(tl.loc('HelmNotFoundInFolder', cachedToolpath));
        }
        fs.chmodSync(helmpath, '777');
        return helmpath;
    });
}
exports.downloadHelm = downloadHelm;
function findHelm(rootFolder) {
    const helmPath = path.join(rootFolder, '*', helmToolName + utility_1.getExecutableExtension());
    const allPaths = tl.find(rootFolder);
    const matchingResultsFiles = tl.match(allPaths, helmPath, rootFolder);
    return matchingResultsFiles[0];
}
function getHelmDownloadURL(version) {
    switch (os.type()) {
        case 'Linux':
            const architecture = osutil.getSupportedLinuxArchitecture();
            return util.format('https://get.helm.sh/helm-%s-linux-%s.zip', version, architecture);
        case 'Darwin':
            return util.format('https://get.helm.sh/helm-%s-darwin-amd64.zip', version);
        case 'Windows_NT':
            return util.format('https://get.helm.sh/helm-%s-windows-amd64.zip', version);
        default:
            throw Error('Unknown OS type');
    }
}
function getStableHelmVersion() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const downloadPath = yield toolLib.downloadTool(helmAllReleasesUrl);
            const responseArray = JSON.parse(fs.readFileSync(downloadPath, 'utf8').toString().trim());
            let latestHelmVersion = semver.clean(stableHelmVersion);
            responseArray.forEach(response => {
                if (response && response.tag_name) {
                    let currentHelmVerison = semver.clean(response.tag_name.toString());
                    if (currentHelmVerison) {
                        if (currentHelmVerison.toString().indexOf('rc') == -1 && semver.gt(currentHelmVerison, latestHelmVersion)) {
                            //If current helm version is not a pre release and is greater than latest helm version
                            latestHelmVersion = currentHelmVerison;
                        }
                    }
                }
            });
            latestHelmVersion = "v" + latestHelmVersion;
            return latestHelmVersion;
        }
        catch (error) {
            let telemetry = {
                event: "HelmLatestNotKnown",
                url: helmAllReleasesUrl,
                error: error
            };
            console.log("##vso[telemetry.publish area=%s;feature=%s]%s", "TaskEndpointId", "HelmInstaller", JSON.stringify(telemetry));
            tl.warning(tl.loc('HelmLatestNotKnown', helmAllReleasesUrl, error, stableHelmVersion));
        }
        return stableHelmVersion;
    });
}
exports.getStableHelmVersion = getStableHelmVersion;
