// azure-pipelines-tasks-kubernetes-common/helm-object-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Helm = void 0;
const tl = require("azure-pipelines-task-lib/task");
class Helm {
    constructor(kubectlPath, namespace) {
        this.helmPath = kubectlPath;
        this.setHelmVersion();
        if (!!namespace) {
            this.namespace = namespace;
        }
        else {
            this.namespace = 'default';
        }
    }
    setHelmVersion() {
        try {
            const result = tl.execSync(this.helmPath, ["version", "--short"], { silent: true });
            this.isHelmV3 = result.stdout.startsWith('v3');
        }
        catch (error) {
            this.isHelmV3 = false;
        }
    }
    template(releaseName, chartPath, overrideFiles, overrideValues) {
        const command = tl.tool(this.helmPath);
        let args = [];
        args.push('template');
        if (!this.isHelmV3) {
            args.push(chartPath);
            if (releaseName) {
                args.push('--name');
                args.push(releaseName);
            }
        }
        else {
            if (releaseName) {
                args.push(releaseName);
            }
            args.push(chartPath);
        }
        args.push('--namespace');
        args.push(this.namespace);
        if (overrideFiles.length > 0) {
            overrideFiles.forEach(file => {
                args.push('-f');
                args.push(file);
            });
        }
        args = args.concat(this._buildSetArguments(overrideValues));
        command.arg(args);
        // Printing the command explicitly because it gets masked when silent is set to true
        console.log(`[command] ${this.helmPath} ${args.join(' ') || ''}`);
        return command.execSync({ silent: true });
    }
    _buildSetArguments(overrideValues) {
        const newArgs = [];
        overrideValues.forEach(overrideValue => {
            newArgs.push('--set');
            newArgs.push(`${overrideValue.name}=${overrideValue.value}`);
        });
        return newArgs;
    }
}
exports.Helm = Helm;
