// azure-pipelines-tasks-kubernetes-common/kubernetesconstants.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.pipelineAnnotations = exports.workloadTypesWithRolloutStatus = exports.workloadTypes = exports.deploymentTypes = exports.ServiceTypes = exports.DiscoveryAndLoadBalancerResource = exports.KubernetesWorkload = void 0;
const tl = require("azure-pipelines-task-lib/task");
const utility_1 = require("./utility");
class KubernetesWorkload {
}
exports.KubernetesWorkload = KubernetesWorkload;
KubernetesWorkload.pod = 'Pod';
KubernetesWorkload.replicaset = 'Replicaset';
KubernetesWorkload.deployment = 'Deployment';
KubernetesWorkload.statefulSet = 'StatefulSet';
KubernetesWorkload.daemonSet = 'DaemonSet';
KubernetesWorkload.job = 'job';
KubernetesWorkload.cronjob = 'cronjob';
class DiscoveryAndLoadBalancerResource {
}
exports.DiscoveryAndLoadBalancerResource = DiscoveryAndLoadBalancerResource;
DiscoveryAndLoadBalancerResource.service = 'service';
DiscoveryAndLoadBalancerResource.ingress = 'ingress';
class ServiceTypes {
}
exports.ServiceTypes = ServiceTypes;
ServiceTypes.loadBalancer = 'LoadBalancer';
ServiceTypes.nodePort = 'NodePort';
ServiceTypes.clusterIP = 'ClusterIP';
exports.deploymentTypes = ['deployment', 'replicaset', 'daemonset', 'pod', 'statefulset'];
exports.workloadTypes = ['deployment', 'replicaset', 'daemonset', 'pod', 'statefulset', 'job', 'cronjob'];
exports.workloadTypesWithRolloutStatus = ['deployment', 'daemonset', 'statefulset'];
const isRelease = utility_1.isEqual(tl.getVariable('SYSTEM_HOSTTYPE'), 'release', true);
const orgUrl = tl.getVariable('System.TeamFoundationCollectionUri');
exports.pipelineAnnotations = [];
if (isRelease) {
    exports.pipelineAnnotations = [
        `azure-pipelines/run=${tl.getVariable('Release.ReleaseId')}`,
        `azure-pipelines/pipeline="${tl.getVariable('Release.DefinitionName')}"`,
        `azure-pipelines/pipelineId="${tl.getVariable('Release.DefinitionId')}"`,
        `azure-pipelines/jobName="${tl.getVariable('Agent.JobName')}"`,
        `azure-pipelines/runuri=${orgUrl}${tl.getVariable('System.TeamProject')}/_releaseProgress?releaseId=${tl.getVariable('Release.ReleaseId')}`,
        `azure-pipelines/project=${tl.getVariable('System.TeamProject')}`,
        `azure-pipelines/org=${orgUrl}`
    ];
}
else {
    exports.pipelineAnnotations = [
        `azure-pipelines/run=${tl.getVariable('Build.BuildNumber')}`,
        `azure-pipelines/pipeline="${tl.getVariable('Build.DefinitionName')}"`,
        `azure-pipelines/pipelineId="${tl.getVariable('System.DefinitionId')}"`,
        `azure-pipelines/jobName="${tl.getVariable('Agent.JobName')}"`,
        `azure-pipelines/runuri=${orgUrl}${tl.getVariable('System.TeamProject')}/_build/results?buildId=${tl.getVariable('Build.BuildId')}`,
        `azure-pipelines/project=${tl.getVariable('System.TeamProject')}`,
        `azure-pipelines/org=${orgUrl}`
    ];
}
