// azure-pipelines-tasks-kubernetes-common/kubectl-object-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Kubectl = void 0;
const tl = require("azure-pipelines-task-lib/task");
const path = require("path");
tl.setResourcePath(path.join(__dirname, 'module.json'), true);
class Kubectl {
    constructor(kubectlPath, namespace, ignoreSSLErrors) {
        this.kubectlPath = kubectlPath;
        this.ignoreSSLErrors = !!ignoreSSLErrors;
        if (!!namespace) {
            this.namespace = namespace;
        }
        else {
            this.namespace = 'default';
        }
        this.displayKubectlVersion();
    }
    apply(configurationPaths) {
        const command = tl.tool(this.kubectlPath);
        command.arg('apply');
        command.arg(['-f', this.createInlineArray(configurationPaths)]);
        return this.execute(command);
    }
    annotate(resourceType, resourceName, annotations, overwrite) {
        const command = tl.tool(this.kubectlPath);
        command.arg('annotate');
        command.arg([resourceType, resourceName]);
        command.arg(annotations);
        if (!!overwrite) {
            command.arg(`--overwrite`);
        }
        return this.execute(command);
    }
    annotateFiles(files, annotations, overwrite) {
        const command = tl.tool(this.kubectlPath);
        command.arg('annotate');
        command.arg(['-f', this.createInlineArray(files)]);
        command.arg(annotations);
        if (!!overwrite) {
            command.arg(`--overwrite`);
        }
        return this.execute(command);
    }
    createDockerSecret(secretName, registryServer, userName, password, email, force) {
        if (!!force && !!secretName) {
            this.deleteSecret(secretName);
        }
        const command = tl.tool(this.kubectlPath);
        command.arg('create');
        command.arg('secret');
        command.arg('docker-registry');
        command.arg(secretName);
        command.arg(['--docker-username', userName]);
        command.arg(['--docker-password', password]);
        command.arg(['--docker-server', registryServer]);
        command.arg(['--docker-email', email]);
        return this.execute(command);
    }
    createGenericSecret(secretName, args, force) {
        if (!!force && !!secretName) {
            this.deleteSecret(secretName);
        }
        const command = tl.tool(this.kubectlPath);
        command.arg('create');
        command.arg('secret');
        command.arg('generic');
        command.arg(secretName);
        if (args) {
            command.line(args);
        }
        return this.execute(command);
    }
    describe(resourceType, resourceName, silent) {
        const command = tl.tool(this.kubectlPath);
        command.arg('describe');
        command.arg([resourceType, resourceName]);
        return this.execute(command, silent);
    }
    getNewReplicaSet(deployment) {
        let newReplicaSet = '';
        const result = this.describe('deployment', deployment, true);
        if (result != null && result.stdout != null) {
            const stdout = result.stdout.split('\n');
            stdout.forEach((line) => {
                if (!!line && line.toLowerCase().indexOf('newreplicaset') > -1) {
                    newReplicaSet = line.substr(14).trim().split(' ')[0];
                }
            });
        }
        return newReplicaSet;
    }
    getAllPods() {
        const command = tl.tool(this.kubectlPath);
        command.arg('get');
        command.arg('pods');
        command.arg(['-o', 'json']);
        return this.execute(command, true);
    }
    getClusterInfo() {
        const command = tl.tool(this.kubectlPath);
        command.arg('cluster-info');
        return this.execute(command, true);
    }
    checkRolloutStatus(resourceType, name, timeoutInSeconds) {
        const command = tl.tool(this.kubectlPath);
        command.arg(['rollout', 'status']);
        command.arg(resourceType + '/' + name);
        if (timeoutInSeconds)
            command.arg(['--timeout', timeoutInSeconds + 's']);
        return this.execute(command);
    }
    getResource(resourceType, name) {
        const command = tl.tool(this.kubectlPath);
        command.arg('get');
        command.arg(resourceType + '/' + name);
        command.arg(['-o', 'json']);
        return this.execute(command);
    }
    getResources(applyOutput, filterResourceTypes) {
        const outputLines = applyOutput.split('\n');
        const results = [];
        outputLines.forEach(line => {
            if (line && line.trim().length > 0) {
                const words = line.split(' ');
                const resourceInfo = words[0].trim().split('/');
                const resourceType = resourceInfo[0];
                const resourceName = resourceInfo[1];
                if (filterResourceTypes.filter(type => !!type && resourceType.toLowerCase().startsWith(type.toLowerCase())).length > 0) {
                    results.push({
                        type: resourceType,
                        name: resourceName
                    });
                }
            }
        });
        return results;
    }
    scale(resourceType, resourceName, replicas) {
        const command = tl.tool(this.kubectlPath);
        command.arg('scale');
        command.arg(resourceType + '/' + resourceName);
        command.arg(`--replicas=${replicas}`);
        return this.execute(command);
    }
    patch(resourceType, resourceName, patch, strategy) {
        const command = tl.tool(this.kubectlPath);
        command.arg('patch');
        command.arg([resourceType, resourceName]);
        command.arg(`--type=${strategy}`);
        command.arg([`-p`, patch]);
        return this.execute(command);
    }
    delete(args) {
        const command = tl.tool(this.kubectlPath);
        command.arg('delete');
        command.line(args);
        return this.execute(command);
    }
    executeCommand(customCommand, args, silent) {
        const command = tl.tool(this.kubectlPath);
        command.arg(customCommand);
        if (args)
            command.line(args);
        return this.execute(command, silent);
    }
    execute(command, silent) {
        if (this.ignoreSSLErrors) {
            command.arg('--insecure-skip-tls-verify');
        }
        command.arg(['--namespace', this.namespace]);
        return command.execSync({ silent: !!silent });
    }
    createInlineArray(str) {
        if (typeof str === 'string') {
            return str;
        }
        return str.join(',');
    }
    deleteSecret(secretName) {
        const command = tl.tool(this.kubectlPath);
        command.arg('delete');
        command.arg('secret');
        command.arg(secretName);
        this.execute(command);
    }
    displayKubectlVersion() {
        try {
            const result = this.executeCommand('version', '-o json', true);
            const resultInJSON = JSON.parse(result.stdout);
            if (resultInJSON.clientVersion && resultInJSON.clientVersion.gitVersion) {
                console.log('==============================================================================');
                console.log('\t\t\t' + tl.loc('KubectlClientVersion') + ': ' + resultInJSON.clientVersion.gitVersion);
                if (resultInJSON.serverVersion && resultInJSON.serverVersion.gitVersion) {
                    console.log('\t\t\t' + tl.loc('KubectlServerVersion') + ': ' + resultInJSON.serverVersion.gitVersion);
                    console.log('==============================================================================');
                }
                else {
                    console.log('\t' + tl.loc('KubectlServerVersion') + ': ' + tl.loc('KubectlServerVerisonNotFound'));
                    console.log('==============================================================================');
                    tl.debug(tl.loc('UnableToFetchKubectlVersion'));
                }
            }
        }
        catch (ex) {
            console.log(tl.loc('UnableToFetchKubectlVersion'));
        }
    }
}
exports.Kubectl = Kubectl;
