// azure-pipelines-tasks-kubernetes-common/kubernetesconnection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KubernetesConnection = void 0;
const tl = require("azure-pipelines-task-lib/task");
const path = require("path");
const fs = require("fs");
const kubectlutility = require("./kubectlutility");
class KubernetesConnection {
    constructor(kubernetesServiceConnection, tempDir) {
        this.kubernetesServiceConnection = kubernetesServiceConnection;
        this.tempDir = tempDir;
    }
    open() {
        let kubeconfig;
        let kubeconfigFile;
        const authorizationType = tl.getEndpointDataParameter(this.kubernetesServiceConnection, 'authorizationType', true);
        if (!authorizationType || authorizationType === 'Kubeconfig') {
            kubeconfig = kubectlutility.getKubeconfigForCluster(this.kubernetesServiceConnection);
        }
        else if (authorizationType === 'ServiceAccount' || authorizationType === 'AzureSubscription') {
            kubeconfig = kubectlutility.createKubeconfig(this.kubernetesServiceConnection);
        }
        kubeconfigFile = path.join(this.tempDir, 'config');
        fs.writeFileSync(kubeconfigFile, kubeconfig);
        tl.setVariable('KUBECONFIG', kubeconfigFile);
        this.ignoreSSLErrors = tl.getEndpointDataParameter(this.kubernetesServiceConnection, 'acceptUntrustedCerts', true) === 'true';
    }
    close() {
        if (tl.getVariable('KUBECONFIG')) {
            tl.setVariable('KUBECONFIG', '');
        }
    }
}
exports.KubernetesConnection = KubernetesConnection;
